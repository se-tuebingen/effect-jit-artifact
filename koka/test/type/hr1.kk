fun choose(x,y) { 
  if (True) then x else y 
}

fun xauto(f:forall<a> a -> a) {
  f(f)
}

fun auto(f:forall<a> a -> e a) : e (forall<a> a -> e a)
{
  f(f)
}

val test1 = auto(id)
val test2 = apply(auto,id)
val ids   = single(id:forall<a> a -> a)  //rigid annotation
val test3  = ids.map(auto)
val test3a = ids.map(xauto)

fun headlist( xs : list<a> ) : exn a
  head(xs).unjust

fun taillist( xs : list<a> ) : list<a>
  tail(xs)  

fun test4()  { apply(headlist,ids) }
fun test4a() { apply(taillist,ids) }

fun test5()  { apply(fn(xs:list<list<forall<a> a -> a>>){ xs.map(headlist) },single(ids)) }

fun test6()  { choose(id,auto) }
fun test6a() { choose(auto,id) }
fun test7()  { choose(xauto,xauto) }

fun test8()  { choose(ids,[]) }
fun test8a() { choose([],ids) }
