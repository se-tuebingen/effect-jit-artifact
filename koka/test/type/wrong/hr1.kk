fun runST( action : forall<h> () -> <st<h>,exn> a ) : exn a  {
  throw("runST")
}

fun wrong() {
  val r = ref(1)
  runST( fn(){ !r } )
}

fun main() {
  print("hi")
}