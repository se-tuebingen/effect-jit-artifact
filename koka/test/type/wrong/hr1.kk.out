test/type/wrong/hr1@kk(7,10): type error: abstract type(s) escape(s) into the context
 term : fn(){ !r }
 inferred type: () -> <st<$h>,exn> int
 hint : give a higher-rank type annotation to a function parameter?